cnd/src/btsieve/ethereum/mod.rs

```diff
@@ -24,16 +20,150 @@ use std::collections::HashSet;
 type Hash = H256;
 type Block = crate::ethereum::Block<Transaction>;
 
-pub async fn matching_transaction<C>(
+pub async fn matching_create_contract_transaction<C>(
+    blockchain_connector: C,
+    start_of_swap: NaiveDateTime,
+    bytecode: Bytes,
+) -> anyhow::Result<Transaction>
+where
+    C: LatestBlock<Block = Option<Block>>
+        + BlockByHash<Block = Option<Block>, BlockHash = Hash>
+        + ReceiptByHash<Receipt = Option<TransactionReceipt>, TransactionHash = Hash>
+        + Clone,
+{
+    matching_transaction(blockchain_connector, start_of_swap, |block| {
+        Box::new({
+            let bytecode = bytecode.clone();
+
+            async move {
+                for transaction in block.transactions.into_iter() {
+                    // transaction.to address is None if, and only if, the transaction
+                    // creates a contract.
+                    if transaction.to.is_none() && transaction.input == bytecode {
+                        return Ok(Some(transaction));
+                    }
+                }
+                Ok(None)
+            }
+            .boxed()
+        })
+    })
+    .await
+}
+
+pub async fn matching_events_transaction<C>(
+    blockchain_connector: C,
+    start_of_swap: NaiveDateTime,
+    events: Vec<Event>,
+    action: &'static str,
+) -> anyhow::Result<(Transaction, TransactionReceipt)>
+where
+    C: LatestBlock<Block = Option<Block>>
+        + BlockByHash<Block = Option<Block>, BlockHash = Hash>
+        + ReceiptByHash<Receipt = Option<TransactionReceipt>, TransactionHash = Hash>
+        + Clone,
+{
+    matching_transaction_receipt(blockchain_connector.clone(), start_of_swap, |block| {
```

Comment:
You could go ahead and pass the `topics` already to `matching_transaction_receipt`. This would allow you to do the bloom filter matching inside and call the closure with a tuple of transaction and receipt.

Reply:
Nice idea, will consider.

Reply:
It would also invalidate the assumption here: https://github.com/comit-network/comit-rs/pull/2039#discussion_r380997282